quinn-proto/src/token_reuse_preventer.rs

```diff
@@ -0,0 +1,218 @@
+//! Limiting clients' ability to reuse tokens from NEW_TOKEN frames
+
+use std::{
+    hash::{Hash as _, Hasher as _},
+    mem::{size_of, swap},
+    time::{Duration, SystemTime, UNIX_EPOCH},
+};
+
+use rustc_hash::{FxHashSet, FxHasher};
+use tracing::warn;
+
+/// Error for when a token may have been reused
+pub struct TokenReuseError;
+
+/// Responsible for limiting clients' ability to reuse tokens from NEW_TOKEN frames
+///
+/// [_RFC 9000 § 8.1.4:_](https://www.rfc-editor.org/rfc/rfc9000.html#section-8.1.4)
+///
+/// > Attackers could replay tokens to use servers as amplifiers in DDoS attacks. To protect
+/// > against such attacks, servers MUST ensure that replay of tokens is prevented or limited.
+/// > Servers SHOULD ensure that tokens sent in Retry packets are only accepted for a short time,
+/// > as they are returned immediately by clients. Tokens that are provided in NEW_TOKEN frames
+/// > (Section 19.7) need to be valid for longer but SHOULD NOT be accepted multiple times.
+/// > Servers are encouraged to allow tokens to be used only once, if possible; tokens MAY include
+/// > additional information about clients to further narrow applicability or reuse.
+pub trait TokenReusePreventer: Send + Sync {
+    /// Called when a client uses a token from a NEW_TOKEN frame
```

Comment:
This should document what the function should do, not (just) where it's used.

Reply:
Done